data/external/downloader.py

```diff
@@ -578,63 +598,62 @@ def _get_cache_path(fname):
+            return json.load(file)
     else:
         return None
 
 
-def _get_roomsearch_xml(url, params, cache_fname):
+def _get_roomsearch_xml(url: str, params: dict, cache_fname: str) -> BeautifulSoup:
     root = _get_xml(url, params, cache_fname)
-    e = root.find('.//instruction[@jsid="raumSucheKontainerID"]')
-    return BeautifulSoup(e.text, "lxml")
+    elem = root.find('.//instruction[@jsid="raumSucheKontainerID"]')
+    return BeautifulSoup(elem.text, "lxml")
 
 
-def _get_xml(url, params, cache_fname):
+def _get_xml(url: str, params: dict, cache_fname: str):
     cache_path = os.path.join(os.path.dirname(__file__), "cache", cache_fname)
     if os.path.exists(cache_path):
         tree = ET.parse(cache_path)
-        root = tree.getroot()
-    else:
-        print(f"GET {url}", params)
-        r = requests.get(url, params)
-        root = ET.fromstring(r.text)
-        with open(cache_path, "w") as f:
-            f.write(r.text)
-    return root
+        return tree.getroot()
 
+    print(f"GET {url}", params)
+    req = requests.get(url, params)
+    with open(cache_path, "w", encoding="utf-8") as file:
+        file.write(req.text)
+    return ET.fromstring(req.text)
 
-def _get_html(url, params, cache_fname):
+
+def _get_html(url: str, params: dict, cache_fname: str) -> BeautifulSoup:
     cache_path = os.path.join(os.path.dirname(__file__), "cache", cache_fname)
     if os.path.exists(cache_path):
-        with open(cache_path) as f:
-            return BeautifulSoup(f.read(), "lxml")
+        with open(cache_path, encoding="utf-8") as file:
+            return BeautifulSoup(file.read(), "lxml")
     else:
-        r = requests.get(url, params)
+        req = requests.get(url, params)
         time.sleep(0.5)  # Not the best place to put this
-        with open(cache_path, "w") as f:
-            f.write(r.text)
-        return BeautifulSoup(r.text, "lxml")
+        with open(cache_path, "w", encoding="utf-8") as file:
+            file.write(req.text)
+        return BeautifulSoup(req.text, "lxml")
 
 
 def _download_file(url, cache_rel_path):
     cache_path = os.path.join(os.path.dirname(__file__), "cache", cache_rel_path)
 
     if not os.path.exists(cache_path):
         print(f"Retrieving: '{url}'")
-        urllib.request.urlretrieve(url, cache_path)
+        # url parameter does not allow path traversal, because we build it further up in the callstack
+        urllib.request.urlretrieve(url, cache_path)  # nosec: B310
 
     return cache_path
 
 
 def _write_cache_json(fname, data):
     path = _get_cache_path(fname)
-    with open(path, "w") as f:
-        json.dump(data, f)
+    with open(path, "w", encoding="utf-8") as file:
+        json.dump(data, file)
 
 
 def _get_tumonline_api_url(base_target):
-    return "https://campus.tum.de/tumonline/{}/NC_{}".format(
-        base_target,
-        str(random.randint(0, 9999)).zfill(4),
-    )
+    # TODO: WTF is this ???
```

Comment:
I have absolutely no idea, observed it in the API

Reply:
```suggestion
    # I have no idea, what this magic_string is, or why it exists.. Usage is the same as from TUMOnline..
```